File: crypto/src/main/java/com/blockstream/gdk/data/AccountType.kt
package com.blockstream.gdk.data


enum class AccountType(val gdkType: String) {
    // Mutlisig
    STANDARD("2of2"),
    AMP_ACCOUNT("2of2_no_recovery"),
    TWO_OF_THREE("2of3"),

    // Singlesig
    BIP44_LEGACY("p2pkh"),
    BIP49_SEGWIT_WRAPPED("p2sh-p2wpkh"),
    BIP84_SEGWIT("p2wpkh"),
    BIP86_TAPROOT("p2tr"),

    LIGHTNING("lightning"),

    UNKNOWN("unknown");

    override fun toString(): String = when(this){
        BIP44_LEGACY -> "Legacy"
        BIP49_SEGWIT_WRAPPED -> "Legacy"
        BIP84_SEGWIT -> "SegWit"
        BIP86_TAPROOT -> "Taproot"
        LIGHTNING -> "Lightning"
        else -> gdkType
    }

    fun isSinglesig() = when (this) {
        BIP44_LEGACY, BIP49_SEGWIT_WRAPPED, BIP84_SEGWIT, BIP86_TAPROOT -> true
        else -> false
    }

    fun isMutlisig() = !isSinglesig()

    companion object {
        fun byGDKType(name: String) = when(name){
            "2of2" -> STANDARD
            "2of2_no_recovery" -> AMP_ACCOUNT
            "2of3" -> TWO_OF_THREE
            "p2pkh" -> BIP44_LEGACY
            "p2sh-p2wpkh" -> BIP49_SEGWIT_WRAPPED
            "p2wpkh" -> BIP84_SEGWIT
            "p2tr" -> BIP86_TAPROOT
            else -> UNKNOWN
        }
    }
}

File: gaios/Accounts/AccountType.swift
enum AccountType: String, CaseIterable, Codable, Comparable {
    /// multiSig
    case standard = "2of2"
    case amp = "2of2_no_recovery"
    case twoOfThree = "2of3"

    /// singleSig
    case legacy = "p2pkh"
    case segwitWrapped = "p2sh-p2wpkh" // former legacy
    case segWit = "p2wpkh"
    case taproot = "p2tr"

    var typeStringId: String {
        get {
            switch self {
            case .standard:
                return "id_standard"
            case .amp:
                return "id_amp"
            case .twoOfThree:
                return "id_2of3"
            case .legacy:
                return "id_legacy_bip44"
            case .segwitWrapped:
                return "id_legacy_segwit_bip49"
            case .segWit:
                return "id_segwit_bip84"
            case .taproot:
                return "id_taproot_bip86"
            }
        }
    }

    var nameStringId: String {
        get {
            switch self {
            case .standard:
                return "id_standard_account"
            case .amp:
                return "id_amp_account"
            case .twoOfThree:
                return "id_2of3_account"
            case .legacy:
                return "id_legacy_segwit_account"
            case .segwitWrapped:
                return "id_legacy_segwit_account"
            case .segWit:
                return "id_segwit_account"
            case .taproot:
                return "id_taproot_account"
            }
        }
    }

    var shortNameStringId: String {
        get {
            switch self {
            case .standard:
                return "id_standard"
            case .amp:
                return "id_amp"
            case .twoOfThree:
                return "id_2of3"
            case .legacy:
                return "Legacy"
            case .segwitWrapped:
                return "Legacy SegWit"
            case .segWit:
                return "SegWit"
            case .taproot:
                return "Taproot"
            }
        }
    }

    static func < (a: AccountType, b: AccountType) -> Bool {
        if a == .segWit {
            return true
        } else if b == .segWit {
            return false
        } else {
           return a.rawValue < b.rawValue
        }
    }
}

enum RecoveryKeyType {
    case hw
    case newPhrase(lenght: Int)
    case existingPhrase
    case publicKey
}
